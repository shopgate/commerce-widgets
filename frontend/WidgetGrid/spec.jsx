/**
 * Copyright (c) 2017, Shopgate, Inc. All rights reserved.
 *
 * This source code is licensed under the Apache 2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

import React from 'react';
import { mount } from 'enzyme';
import extensions from 'Extensions';
import WidgetGrid from './index';

describe('<WidgetGrid />', () => {
  it('should render with a config', () => {
    const config = [{
      col: 0,
      row: 0,
      width: 12,
      height: 3,
      settings: {
        id: 83535,
        image: 'https://data.shopgate.com/shop_widget_images/23836/92204c0f264ac30d6836994c2fb64eb1.min.jpeg',
      },
      type: 'core-widgets/image',
    }];

    const wrapper = mount(
      <WidgetGrid config={config} />
    );

    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find(extensions['core-widgets/image']).exists()).toBe(true);
  });

  it('should not render without a `config` prop', () => {
    const wrapper = mount(

      <WidgetGrid />
    );

    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find('div').children().length).toBe(0);
  });
});
