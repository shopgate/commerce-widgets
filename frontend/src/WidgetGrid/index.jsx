/**
 * Copyright (c) 2017, Shopgate, Inc. All rights reserved.
 *
 * This source code is licensed under the Apache 2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

import React from 'react';
import PropTypes from 'prop-types';
import sortBy from 'lodash/sortBy';
import extensions from 'Extensions';
import Widget from 'Components/Widgets/components/Widget';
import styles from './style';

// One grid row has 12 columns.
const GRID_COLUMNS = 12;

/**
 * Iterate through all widgets and return the maxium
 * height based on row and height information.
 * @param {Array} widgets Array of widgets.
 * @returns {number} Height of the widget grid.
 */
const getMaxHeight = widgets => (
  widgets.reduce(
    (max, widget) => Math.max(widget.row + widget.height, max),
    0
  )
);

/**
 * The widget grid widget component.
 * @param {Object} props The component properties.
 * @returns {JSX} The widget grid.
 */
const WidgetGrid = (props) => {
  const rowCount = getMaxHeight(props.config);

  // The cell size is 1/12 of the viewport width.
  const cellSize = Math.floor(window.innerWidth / GRID_COLUMNS);

  if (!props.config || !rowCount) {
    return null;
  }

  // Sort the widgets by row. This has to happen to take care of the z-index flow.
  const widgets = sortBy(props.config, ['row']);

  // The height of the widget area.
  const height = `${rowCount * cellSize}px`;

  return (
    <div className={styles.container} style={{ height }}>
      {Object.keys(widgets).map((key) => {
        const widget = widgets[key];
        const widgetKey = `w${key}`;
        // Map to the correct widget component using the `type` key inside the widget.
        const WidgetComponent = extensions[widget.type];

        return (
          <Widget
            cellSize={cellSize}
            config={widget}
            component={WidgetComponent}
            key={widgetKey}
          />
        );
      })}
    </div>
  );
};

WidgetGrid.propTypes = {
  config: PropTypes.arrayOf(PropTypes.shape()),
};

WidgetGrid.defaultProps = {
  config: [],
};

export default WidgetGrid;
