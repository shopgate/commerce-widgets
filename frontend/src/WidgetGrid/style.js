/**
 * Copyright (c) 2017, Shopgate, Inc. All rights reserved.
 *
 * This source code is licensed under the Apache 2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

import css from 'glamor';
import colors from 'Styles/colors';

const container = css({
  background: colors.background,
  position: 'relative',
}).toString();

export default {
  container,
};
