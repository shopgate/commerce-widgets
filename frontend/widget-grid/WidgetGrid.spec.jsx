import React from 'react';
import { mount } from 'enzyme';
import extensions from 'Extensions';
import WidgetGrid from './WidgetGrid';

describe('<WidgetGrid />', () => {
  it('should render with a config', () => {
    const config = [{
      col: 0,
      row: 0,
      width: 12,
      height: 3,
      settings: {
        id: 83535,
        image: 'https://data.shopgate.com/shop_widget_images/23836/92204c0f264ac30d6836994c2fb64eb1.min.jpeg',
      },
      type: 'core-widgets/image',
    }];

    const wrapper = mount(
      <WidgetGrid config={config} />
    );

    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find(extensions['core-widgets/image']).exists()).toBe(true);
  });

  it('should not render without a `config` prop', () => {
    const wrapper = mount(

      <WidgetGrid />
    );

    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find('div').children().length).toBe(0);
  });
});
