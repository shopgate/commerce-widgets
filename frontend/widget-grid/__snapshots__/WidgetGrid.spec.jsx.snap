// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<WidgetGrid /> should not render without a \`config\` prop 1`] = `
<WidgetGrid
  config={Array []}
/>
`;

exports[`<WidgetGrid /> should render with a config 1`] = `
<WidgetGrid
  config={
    Array [
      Object {
        "col": 0,
        "height": 3,
        "row": 0,
        "settings": Object {
          "id": 83535,
          "image": "https://data.shopgate.com/shop_widget_images/23836/92204c0f264ac30d6836994c2fb64eb1.min.jpeg",
        },
        "type": "core-widgets/image",
        "width": 12,
      },
    ]
  }
>
  <div
    className="b-f8f8f8 position-relative"
    style={
      Object {
        "height": "255px",
      }
    }
  >
    <Widget
      cellSize={85}
      component={[Function]}
      config={
        Object {
          "col": 0,
          "height": 3,
          "row": 0,
          "settings": Object {
            "id": 83535,
            "image": "https://data.shopgate.com/shop_widget_images/23836/92204c0f264ac30d6836994c2fb64eb1.min.jpeg",
          },
          "type": "core-widgets/image",
          "width": 12,
        }
      }
    >
      <GridItem
        className="d-block position-absolute h-255px w-100P  "
        component="div"
        grow={0}
        shrink={1}
      >
        <div
          className="d-block position-absolute h-255px w-100P  "
        >
          <div
            className="position-absolute top-0 left-0 right-0 bottom-0 overflow-hidden"
          >
            <ImageWidget
              ratio={
                Array [
                  12,
                  3,
                ]
              }
              settings={
                Object {
                  "id": 83535,
                  "image": "https://data.shopgate.com/shop_widget_images/23836/92204c0f264ac30d6836994c2fb64eb1.min.jpeg",
                }
              }
            >
              <Image
                alt={null}
                animating={true}
                className=""
                forcePlaceholder={false}
                highestResolutionLoaded={[Function]}
                onError={null}
                onLoad={null}
                ratio={
                  Array [
                    12,
                    3,
                  ]
                }
                resolutions={
                  Array [
                    Object {
                      "blur": 2,
                      "height": 50,
                      "width": 50,
                    },
                    Object {
                      "height": 440,
                      "width": 440,
                    },
                  ]
                }
                src="https://data.shopgate.com/shop_widget_images/23836/92204c0f264ac30d6836994c2fb64eb1.min.jpeg"
              >
                <Transition
                  childrenStyles={
                    Object {
                      "appear": Object {
                        "opacity": 1,
                        "transition": "opacity 300ms cubic-bezier(0.4, 0.0, 0.2, 1)",
                      },
                      "base": Object {
                        "opacity": 0,
                        "transition": "opacity 300ms cubic-bezier(0.4, 0.0, 0.2, 1)",
                      },
                      "enter": Object {
                        "opacity": 1,
                        "transition": "opacity 300ms cubic-bezier(0.4, 0.0, 0.2, 1)",
                      },
                      "leave": Object {
                        "opacity": 0,
                        "transition": "opacity 300ms cubic-bezier(0.4, 0.0, 0.2, 1)",
                      },
                    }
                  }
                  className="position-relative b-f2f2f2 z-index-0 _before-d-block _before-content-__ _before-w-100P _before-pt-25p000P "
                  component="div"
                >
                  <TransitionHooks
                    className="position-relative b-f2f2f2 z-index-0 _before-d-block _before-content-__ _before-w-100P _before-pt-25p000P "
                    component="div"
                  >
                    <div
                      className="position-relative b-f2f2f2 z-index-0 _before-d-block _before-content-__ _before-w-100P _before-pt-25p000P "
                    />
                  </TransitionHooks>
                </Transition>
              </Image>
            </ImageWidget>
          </div>
        </div>
      </GridItem>
    </Widget>
  </div>
</WidgetGrid>
`;
